velox/row/UnsafeRowFast.cpp

```diff
@@ -0,0 +1,369 @@
+
+  int32_t totalSize = kFieldWidth;
+
+  auto keysSize = serializeAsArray(
+      children_[0], offset, size, fixedWidth_[0], buffer + totalSize);
+  totalSize += keysSize;
+
+  auto valuesSize = serializeAsArray(
+      children_[1], offset, size, fixedWidth_[1], buffer + totalSize);
+  totalSize += valuesSize;
+
+  // Write the size of serialized keys.
+  *reinterpret_cast<int64_t*>(buffer) = keysSize;
+
+  return totalSize;
+}
+
+int32_t UnsafeRowFast::arrayRowSize(
+    UnsafeRowFast& elements,
+    vector_size_t offset,
+    vector_size_t size,
+    bool fixedWidth) {
+  int32_t nullBytes = alignBits(size);
+
+  int32_t rowSize = kFieldWidth + nullBytes;
+  if (fixedWidth) {
+    return rowSize + size * elements.valueBytes();
+  }
+
+  rowSize += size * kFieldWidth;
+
+  for (auto i = 0; i < size; ++i) {
+    if (!elements.isNullAt(offset + i)) {
+      rowSize += alignBytes(elements.variableWidthRowSize(offset + i));
+    }
+  }
+
+  return rowSize;
+}
+
+int32_t UnsafeRowFast::serializeAsArray(
+    UnsafeRowFast& elements,
+    vector_size_t offset,
+    vector_size_t size,
+    bool fixedWidth,
+    char* buffer) {
+  // array size | null bits | fixed-width data | variable-width data
+
+  // Write array size.
+  *reinterpret_cast<int64_t*>(buffer) = size;
+
+  int32_t nullBytes = alignBits(size);
+
+  int32_t nullsOffset = sizeof(int64_t);
+  int32_t fixedWidthOffset = nullsOffset + nullBytes;
+
+  auto childSize = fixedWidth ? elements.valueBytes() : kFieldWidth;
+
+  int64_t variableWidthOffset = fixedWidthOffset + size * childSize;
+
+  if (elements.supportsBulkCopy_) {
+    if (elements.decoded_.mayHaveNulls()) {
+      for (auto i = 0; i < size; ++i) {
+        if (elements.isNullAt(offset + i)) {
+          bits::setBit(buffer + nullsOffset, i, true);
+        }
+      }
+    }
+    elements.serializeFixedWidth(offset, size, buffer + fixedWidthOffset);
+    return variableWidthOffset;
+  }
+
+  for (auto i = 0; i < size; ++i) {
+    if (elements.isNullAt(offset + i)) {
+      bits::setBit(buffer + nullsOffset, i, true);
+    } else {
+      if (fixedWidth) {
+        elements.serializeFixedWidth(
+            offset + i, buffer + fixedWidthOffset + i * childSize);
+      } else {
+        auto size = elements.serializeVariableWidth(
+            offset + i, buffer + variableWidthOffset);
+
+        // Write size and offset.
+        uint64_t sizeAndOffset = variableWidthOffset << 32 | size;
+        reinterpret_cast<uint64_t*>(buffer + fixedWidthOffset)[i] =
+            sizeAndOffset;
+
+        variableWidthOffset += alignBytes(size);
+      }
+    }
+  }
+  return variableWidthOffset;
+}
+
+int32_t UnsafeRowFast::rowRowSize(vector_size_t index) {
+  auto childIndex = decoded_.index(index);
+
+  const auto numFields = children_.size();
+  int32_t size = rowNullBytes_ + numFields * kFieldWidth;
+
+  for (auto i = 0; i < numFields; ++i) {
+    if (!fixedWidth_[i] && !children_[i].isNullAt(childIndex)) {
```

Comment:
Do we need to include variable width child size here for row size calculation? Thanks!

Reply:
Since the value is null, variable width will be zero. Hence, no need to add that.